File: cpp/api/src/coms/impl/zmq/SubscriberZmq.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include "SubscriberZmq.h"

#include "This.h"
#include "Messages.h"
#include "Waiting.h"
#include "RequestSocket.h"
#include "ContextZmq.h"
#include "IdGenerator.h"
#include "../PublisherImpl.h"
#include "../../../base/JSON.h"

namespace cameo {
namespace coms {

SubscriberZmq::SubscriberZmq() :
	m_appId{0},
	m_pollingTime{100},
	m_timeout{0},
	m_ended{false},
	m_canceled{false},
	m_timedout{false} {
}

SubscriberZmq::~SubscriberZmq() {
}

void SubscriberZmq::setPollingTime(int value) {
	m_pollingTime = value;
}

void SubscriberZmq::setTimeout(int value) {
	m_timeout = value;
}

void SubscriberZmq::init(int appId, const Endpoint& endpoint, const Endpoint& appStatusEndpoint, const std::string& publisherIdentity, bool checkApp) {

	m_publisherIdentity = publisherIdentity;
	m_appId = appId;
	m_ended = false;
	m_canceled = false;

	// Create a socket for publishing.
	ContextZmq * contextImpl {dynamic_cast<ContextZmq *>(This::getCom().getContext())};
	m_subscriber.reset(new zmq::socket_t{contextImpl->getContext(), zmq::socket_type::sub});
	m_subscriber->connect(endpoint.toString());

	m_subscriber->set(zmq::sockopt::subscribe, publisherIdentity);

	// First define the cancel endpoint.
	m_cancelEndpoint = std::string{"inproc://" + IdGenerator::newStringId()};

	m_cancelPublisher = std::unique_ptr<zmq::socket_t>{new zmq::socket_t{contextImpl->getContext(), zmq::socket_type::pub}};
	m_cancelPublisher->bind(m_cancelEndpoint);

	m_subscriber->connect(m_cancelEndpoint);
	m_subscriber->set(zmq::sockopt::subscribe, message::Event::CANCEL);

	// Connect the status publisher if the app is checked.
	if (checkApp) {
		m_subscriber->connect(appStatusEndpoint.toString().c_str());
		m_subscriber->set(zmq::sockopt::subscribe, message::Event::STATUS);
	}

	// Set the poll item.
	m_items[0].socket = static_cast<void *>(*(m_subscriber.get()));
	m_items[0].fd = 0;
	m_items[0].events = ZMQ_POLLIN;
	m_items[0].revents = 0;
}

bool SubscriberZmq::sync(int timeout) {

	// Wait for timeout ms.
	int rc = zmq::poll(m_items, 1, std::chrono::milliseconds{timeout});

	// Return true if the subscriber received a message.
	return (rc != 0);
}

bool SubscriberZmq::hasEnded() const {
	return m_ended;
}

bool SubscriberZmq::isCanceled() const {
	return m_canceled;
}

bool SubscriberZmq::hasTimedout() {
	return m_timedout;
}

bool SubscriberZmq::receiveMessage(zmq::message_t& message) {

	// Define the number of iterations.
	int n {0};
	if (m_pollingTime > 0) {
		n = m_timeout / m_pollingTime + 1;
	}

	// Create the poller.
	zmq::pollitem_t items[] = {
		{ *m_subscriber.get(), 0, ZMQ_POLLIN, 0 }
	};

	// Infinite loop if timeout is 0 or finite loop if timeout is defined.
	int i {0};
	while (i < n || m_timeout == 0) {

		// Check if the requester has been canceled.
		if (m_canceled) {
			return false;
		}

		// Poll the requester.
		zmq::poll(&items[0], 1, std::chrono::milliseconds{m_pollingTime});

		// Get a reply.
		if (items[0].revents & ZMQ_POLLIN) {
			if (!m_subscriber->recv(message, zmq::recv_flags::none).has_value()) {
				return false;
			}
			return true;
		}

		i++;
	}

	// Timeout.
	m_timedout.store(true);

	// No need to reset the socket after a timeout.

	return false;
}

std::optional<std::string> SubscriberZmq::receive() {

	// Reset timedout.
	m_timedout.store(false);

	while (true) {
		zmq::message_t firstPart;
		if (!receiveMessage(firstPart)) {
			return {};
		}

		std::string first {static_cast<char*>(firstPart.data()), firstPart.size()};

		if (first == m_publisherIdentity) {

			zmq::message_t typePart;
			if (!receiveMessage(typePart)) {
				return {};
			}

			std::string messageType {static_cast<char*>(typePart.data()), typePart.size()};

			// Get the JSON object.
			json::Object jsonType;
			json::parse(jsonType, messageType);

			int type {jsonType[message::TYPE].GetInt()};

			if (type == message::STREAM) {
				zmq::message_t dataPart;
				if (!receiveMessage(dataPart)) {
					return {};
				}
				return std::string {static_cast<char*>(dataPart.data()), dataPart.size()};
			}
			else if (type == message::SYNC_STREAM) {
				// Do nothing.
			}
			else if (type == message::STREAM_END) {
				m_ended = true;
				return {};
			}
		}
		else if (first == message::Event::CANCEL) {
			return {};
		}
		else if (first == message::Event::STATUS) {
			zmq::message_t statusPart;
			if (!receiveMessage(statusPart)) {
				return {};
			}

			// Get the JSON object.
			json::Object status;
			json::parse(status, statusPart);

			int id {status[message::StatusEvent::ID].GetInt()};

			if (id == m_appId) {
				state::Value state {status[message::StatusEvent::APPLICATION_STATE].GetInt()};

				// test the terminal state
				if (state == state::SUCCESS
					|| state == state::STOPPED
					|| state == state::KILLED
					|| state == state::FAILURE) {
					// Exit because the remote application has terminated.
					return {};
				}
			}
		}
	}

	return {};
}

std::optional<std::tuple<std::string, std::string>> SubscriberZmq::receiveTwoParts() {

	while (true) {
		zmq::message_t firstPart;
		if (!receiveMessage(firstPart)) {
			return {};
		}

		std::string first {static_cast<char*>(firstPart.data()), firstPart.size()};

		if (first == m_publisherIdentity) {

			zmq::message_t typePart;
			if (!receiveMessage(typePart)) {
				return {};
			}

			std::string messageType {static_cast<char*>(typePart.data()), typePart.size()};

			// Get the JSON object.
			json::Object jsonType;
			json::parse(jsonType, messageType);

			int type {jsonType[message::TYPE].GetInt()};

			if (type == message::STREAM) {

				std::tuple<std::string, std::string> result;

				zmq::message_t data1Part;
				if (!receiveMessage(data1Part)) {
					return {};
				}
				std::string data1 {static_cast<char*>(data1Part.data()), data1Part.size()};

				zmq::message_t data2Part;
				if (!receiveMessage(data2Part)) {
					return {};
				}
				std::string data2 {static_cast<char*>(data2Part.data()), data2Part.size()};

				return std::make_tuple(data1, data2);
			}
			else if (type == message::STREAM_END) {
				m_ended = true;
				return {};
			}
		}
		else if (first == message::Event::CANCEL) {
			return {};
		}
		else if (first == message::Event::STATUS) {
			zmq::message_t statusPart;
			if (!receiveMessage(statusPart)) {
				return {};
			}

			// Get the JSON object.
			json::Object status;
			json::parse(status, statusPart);

			int id {status[message::StatusEvent::ID].GetInt()};

			if (id == m_appId) {
				state::Value state {status[message::StatusEvent::APPLICATION_STATE].GetInt()};

				// test the terminal state
				if (state == state::SUCCESS
					|| state == state::STOPPED
					|| state == state::KILLED
					|| state == state::FAILURE) {
					// Exit because the remote application has terminated.
					return {};
				}
			}
		}
	}

	return {};
}

void SubscriberZmq::cancel() {

	m_canceled = true;

	if (m_cancelPublisher) {
		std::string messageType {message::Event::CANCEL};
		zmq::message_t typePart {messageType.c_str(), messageType.length()};
		m_cancelPublisher->send(typePart, zmq::send_flags::sndmore);

		std::string data {message::Event::CANCEL};
		zmq::message_t dataPart {data.c_str(), data.length()};
		m_cancelPublisher->send(dataPart, zmq::send_flags::none);
	}
}

}
}

File: java/api/src/main/java/eu/ill/cameo/api/coms/impl/zmq/SubscriberZmq.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package eu.ill.cameo.api.coms.impl.zmq;

import java.util.concurrent.atomic.AtomicBoolean;

import org.json.simple.JSONObject;

import eu.ill.cameo.api.base.IdGenerator;
import eu.ill.cameo.api.base.State;
import eu.ill.cameo.api.base.This;
import eu.ill.cameo.api.base.impl.zmq.ContextZmq;
import eu.ill.cameo.api.coms.impl.SubscriberImpl;
import eu.ill.cameo.com.Zmq;
import eu.ill.cameo.common.messages.JSON;
import eu.ill.cameo.common.messages.Messages;
import eu.ill.cameo.common.strings.Endpoint;

public class SubscriberZmq implements SubscriberImpl {
	
	private int pollingTime = 100;
	private int timeout = 0;
	
	private Zmq.Context context;
	private Zmq.Socket subscriber;
	private String publisherIdentity;
	private String cancelEndpoint;
	private Zmq.Socket cancelPublisher;
	private int appId;
	private AtomicBoolean ended = new AtomicBoolean(false);
	private AtomicBoolean canceled = new AtomicBoolean(false);
	private AtomicBoolean timedout = new AtomicBoolean(false);
	private Zmq.Poller poller;
	
	public void setPollingTime(int value) {
		pollingTime = value;
	}
	
	public void setTimeout(int value) {
		timeout = value;
	}
	
	public void init(int appId, Endpoint endpoint, Endpoint appStatusEndpoint, String publisherIdentity, boolean checkApp) {

		this.appId = appId;
		this.publisherIdentity = publisherIdentity;
		
		this.context = ((ContextZmq)This.getCom().getContext()).getContext();
		
		// Create the subscriber
		subscriber = context.createSocket(Zmq.SUB);
		subscriber.connect(endpoint.toString());
		
		// Subcribe to the publisher.
		subscriber.subscribe(publisherIdentity);
		
		// Create an endpoint that should be unique
		cancelEndpoint = "inproc://" + IdGenerator.newStringId();
		
		// Create a cancel publisher so that it sends the CANCEL message to the status subscriber (connected to 2 publishers)
		cancelPublisher = context.createSocket(Zmq.PUB);
		cancelPublisher.bind(cancelEndpoint);

		// Subscribe to CANCEL
		subscriber.connect(cancelEndpoint);
		subscriber.subscribe(Messages.Event.CANCEL);
		
		// Subscribe to STATUS if the app is checked.
		if (checkApp) {
			subscriber.connect(appStatusEndpoint.toString());
			subscriber.subscribe(Messages.Event.STATUS);
		}
		
		// Create the poller.
		poller = this.context.createPoller(1);
		poller.register(subscriber);
	}
	
	public boolean sync(int timeout) {
		
		poller.poll(timeout);
		return poller.pollin(0);
	}
	
	public boolean hasEnded() {
		return ended.get();
	}
	
	public boolean isCanceled() {
		return canceled.get();
	}
	
	@Override
	public boolean hasTimedout() {
		return timedout.get();
	}
	
	private byte[] receiveMessage() {
		
		// Reset timeout.
		timedout.set(false);
		
		// Define the number of iterations.
		int n = 0;
		if (pollingTime > 0) {
			n = timeout / pollingTime + 1;
		}

		// Create the poller.
		Zmq.Poller poller = this.context.createPoller(1);
		poller.register(subscriber);
		
		// Infinite loop if timeout is 0 or finite loop if timeout is defined.
		int i = 0;
		while (i < n || timeout == 0) {

			// Check if the requester has been canceled.
			if (canceled.get()) {
				return null;
			}

			// Poll the requester.
			poller.poll(pollingTime);
			if (poller.pollin(0)) {
				//return Zmq.Msg.recvMsg(subscriber);
				return subscriber.recv();
			}

			i++;
		}

		// Timeout occurred.
		timedout.set(true);

		// Reset the socket because it cannot be reused after a timeout.
		//resetSocket();
		
		return null;
	}
	
	private String receiveStringMessage() {
		
		byte[] response = receiveMessage();
		
		if (response != null) {
			return Messages.parseString(response);
		}
		
		return null;
	}
	
	/**
	 * 
	 * @return the byte[] data. If the return value is null, then the stream is finished. 
	 */
	public byte[] receive() {

		while (true) {
			String message = receiveStringMessage();
			if (message == null) {
				return null;
			}
			
			if (message.equals(publisherIdentity)) {
				
				String messageTypePart = receiveStringMessage();
				if (messageTypePart == null) {
					return null;
				}
				
				// Get the JSON object.
				JSONObject messageType = This.getCom().parse(messageTypePart);
				
				// Get the type.
				long type = JSON.getLong(messageType, Messages.TYPE);
				
				if (type == Messages.STREAM) {
					return receiveMessage();
				}
				else if (type == Messages.SYNC_STREAM) {
					// Do nothing.
				}
				else if (type == Messages.STREAM_END) {
					ended.set(true);
					return null;	
				}
			}
			else if (message.equals(Messages.Event.CANCEL)) {
				return null;
			}
			else if (message.equals(Messages.Event.STATUS)) {
				byte[] statusMessage = receiveMessage();
				if (statusMessage == null) {
					return null;
				}
				
				// Get the JSON object.
				JSONObject status = This.getCom().parse(statusMessage);
				
				// Get the id.
				int id = JSON.getInt(status, Messages.StatusEvent.ID);
									
				if (appId == id) {
					
					// Get the state.
					int state = JSON.getInt(status, Messages.StatusEvent.APPLICATION_STATE);
					
					// Test if the state is terminal
					if (state == State.SUCCESS 
							|| state == State.STOPPED
							|| state == State.KILLED
							|| state == State.FAILURE) {
						// Exit because the remote application has terminated.
						return null;
					}
				}
			}
		}
	}
	
	/**
	 * 
	 * @return the byte[] data. If the return value is null, then the stream is finished. 
	 */
	public byte[][] receiveTwoParts() {

		while (true) {
			String message = receiveStringMessage();
			if (message == null) {
				return null;
			}
			
			if (message.equals(publisherIdentity)) {
				
				String messageTypePart = receiveStringMessage();
				if (messageTypePart == null) {
					return null;
				}
				
				// Get the JSON object.
				JSONObject messageType = This.getCom().parse(messageTypePart);
				
				// Get the type.
				long type = JSON.getLong(messageType, Messages.TYPE);
				
				if (type == Messages.STREAM) {
					byte[][] result = new byte[2][];
					result[0] = receiveMessage();
					result[1] = receiveMessage();
					
					return result;
				}
				else if (type == Messages.STREAM_END) {
					ended.set(true);
					return null;	
				}
			}
			else if (message.equals(Messages.Event.CANCEL)) {
				return null;
			}
			else if (message.equals(Messages.Event.STATUS)) {
				byte[] statusMessage = receiveMessage();
				if (statusMessage == null) {
					return null;
				}
				
				// Get the JSON request object.
				JSONObject request =  This.getCom().parse(statusMessage);
				
				// Get the id.
				int id = JSON.getInt(request, Messages.StatusEvent.ID);
				
				if (appId == id) {
					
					// Get the state.
					int state = JSON.getInt(request, Messages.StatusEvent.APPLICATION_STATE);
					
					// Test if the state is terminal
					if (state == State.SUCCESS 
							|| state == State.STOPPED
							|| state == State.KILLED
							|| state == State.FAILURE) {
						// Exit because the remote application has terminated.
						return null;
					}
				}
			}
		}
	}
	
	/**
	 * 
	 * @return the string data. If the return value is null, then the stream is finished. 
	 */
	public String receiveString() {
		
		byte[] data = receive();
		
		if (data == null) {
			return null;
		}
		
		return Messages.parseString(data);
	}
	
	public void cancel() {	
	
		canceled.set(true);
		
		if (cancelPublisher != null) {
			cancelPublisher.sendMore(Messages.Event.CANCEL);
			cancelPublisher.send(Messages.Event.CANCEL);
		}
	}
	
	public void terminate() {
		
		if (context != null) {
			context.destroySocket(subscriber);
			context.destroySocket(cancelPublisher);
		}
	}
	
}